### ISSUE TITLE
CNTK - Overcoming catastrophic forgetting in neural networks

### ISSUE THREAD

**Issue by pmfcdb** [Microsoft/CNTK#1677]
Hello, is it possible in CNTK to implement this article?

Overcoming catastrophic forgetting in neural networks
http://www.pnas.org/content/114/13/3521.abstract

This could be used to enhanced several domains, such as image recognition , detection, etc... My idea was to improve on existing neural networks models with this technique.

Other idea was to use OpenAI - Universe platform ( https://blog.openai.com/universe/ ) with CNTK to automatic generate input data for training. Is there some work in this subject?

Thks,
Paulo

**Comment by n17s** [comment]
EWC from the paper (especially with diagonal approximation to Fisher Information Matrix) is possible to implement in CNTK with the existing functionality.  Are you interested in implementing it?

**Comment by pmfcdb** [comment]
If it can be implemented with Brainscript and C# Eval wrapper, I would be interested. 

Here is the link of article in pdf : https://arxiv.org/pdf/1612.00796.pdf

It would be really good, if we could have a c# API for CNTK, with the same functions as Brainscript or pythonAPI.

**Comment by n17s** [comment]
It cannot be implemented with either Brainscript of C# Eval. C# Training is in the works. See here: #2194 

To do EWC, take a trained network, compute the gradient on each sample, and sum up the (elementwise) squares of all gradients across the data. This gives the diagonal approximation to the Fisher information matrix. You will then clone the network and create a function that takes the difference of squares between the old and new parameters, scales them by the corresponding sum of squared gradients via elementwise multiplication and finally sums everything up across all parameters. Then you just include that function in your objective with a regularizer hyperparameter that you will have to tune for your task.

Closing this thread. Feel free to reopen if you have further questions.
